### DISCUSSION TITLE
Use DMA with SPI

### DISCUSSION THREAD

**Discussion by PierreGS** [micropython/micropython#10414] [Using MicroPython]
Hello,

I'm working in mircropython on an ESP32-WROVER-E, and I'm looking to run the SPI with the DMA. I don't know how to do it. I have the idea to use registers directly but I'm not sure about this.

Do you have any solution ?

Thank you

Pierre

**Comment by cve2022** [comment]
For direct access to the internal registers, you may use the machine.mem* methods:
[https://docs.micropython.org/en/latest/esp32/tutorial/peripheral_access.html](https://docs.micropython.org/en/latest/esp32/tutorial/peripheral_access.html)
[https://docs.micropython.org/en/latest/library/machine.html](https://docs.micropython.org/en/latest/library/machine.html)

**Reply by stephanelsmith** [reply]
Did not know this was a thing, thank you!

**Comment by PierreGS** [comment]
Thank you. And do you think it's possible to configure DMA with SPI by modify registers ?

**Reply by robert-hh** [reply]
I do not know why you want to do that, since hw SPI at the ESP32 port uses already DMA whenever possible (machine_hw_spi.c, lines 269 ctd.). And your code has anyhow to take care that buffers are valid as long as the transfer takes.

**Reply by PierreGS** [reply]
I saw that the SPI is not using DMA in micropython, by working on it. I'd like to run a web server and a LCD at the same time, with no disturbance between them.

**Reply by robert-hh** [reply]
> I saw that the SPI is not using DMA in micropython

How did you come to that conclusion, at which part of the code did you look?

**Reply by fivdi** [reply]
Assuming that hardware SPI is being used, it is also my understanding that DMA is used for SPI transfers. That being said, the transfers are not asynchronous transfers that occur in the background allowing the calling MicroPython code to continue execution at the same time as the SPI transfer. They are synchronous transfers and the SPI driver waits for the transfers to complete before returning to the calling MicroPython code allowing it to continue execution.

**Reply by robert-hh** [reply]
>  They are synchronous transfers and the SPI driver waits for the transfers to complete before returning to the calling MicroPython code allowing it to continue execution.

Indeed. One of the reasons is, that the data buffers are only guaranteed to be valid during the execution of the call. Asynchronous transfers would require some kind of signalling mechanism for telling if a buffer is still used by the DMA transfer.

**Reply by fivdi** [reply]
I mentioned the synchronous nature of SPI transfers as this may be what is causing issues for @PierreGS. Depending on how often the LCD needs to be updated and how much of the screen is being updated, the update with SPI may be fairly time consuming and block the Web server. However, I'm speculating here.
